src/coreComponents/linearAlgebra/solvers/PreconditionerJacobi.hpp

```diff
@@ -0,0 +1,137 @@
+{
+public:
+
+  /// Alias for base type
+  using Base = PreconditionerBase< LAI >;
+
+  /// Alias for vector type
+  using Vector = typename Base::Vector;
+
+  /// Alias for matrix type
+  using Matrix = typename Base::Matrix;
+
+  virtual ~PreconditionerJacobi()
+  {
+    delete m_diagInv;
+  }
+
+  /**
+   * @brief Compute the preconditioner from a matrix.
+   * @param mat the matrix to precondition.
+   */
+  virtual void compute( Matrix const & mat ) override
+  {
+    GEOSX_LAI_ASSERT( mat.ready() );
+    m_diagInv = new Vector();
+    m_diagInv->createWithLocalSize( mat.numLocalRows(), mat.getComm() );
+    mat.extractDiagonal( *m_diagInv );
+    m_diagInv->reciprocal();
+  }
+
+  /**
+   * @brief Compute the preconditioner from a matrix
+   * @param mat the matrix to precondition
+   * @param dofManager the Degree-of-Freedom manager associated with matrix
+   */
+  virtual void compute( Matrix const & mat,
+                        DofManager const & dofManager ) override
+  {
+    GEOSX_UNUSED_VAR( dofManager );
+    compute( mat );
+  }
+
+  /**
+   * @brief Clean up the preconditioner setup.
+   *
+   * Releases memory used and allows the matrix to be deleted cleanly.
+   * This method should be called before the matrix used to compute the preconditioner
+   * goes out of scope or is re-created. Some implementations require the matrix
+   * to outlive the preconditioner (for example, Trilinos/ML may crash the program if
+   * deleted after the matrix).
+   *
+   * @note Should be properly overridden in derived classes, which may call this method.
+   */
+  virtual void clear() override
+  {
+    m_diagInv->reset();
+  }
+
+  /**
+   * @brief Get the number of global rows.
+   * @return Number of global rows in the operator.
+   */
+  virtual globalIndex numGlobalRows() const override final
+  {
+    return m_diagInv->globalSize();
+  }
+
+  /**
+   * @brief Get the number of global columns.
+   * @return Number of global columns in the operator.
+   */
+  virtual globalIndex numGlobalCols() const override final
+  {
+    return m_diagInv->globalSize();
+  }
+
+  /**
+   * @brief Apply operator to a vector.
+   *
+   * @param src Input vector (src).
+   * @param dst Output vector (dst).
+   */
+  virtual void apply( Vector const & src,
+                      Vector & dst ) const override
+  {
+    GEOSX_LAI_ASSERT_EQ( this->numGlobalRows(), dst.globalSize() );
+    GEOSX_LAI_ASSERT_EQ( this->numGlobalCols(), src.globalSize() );
+    GEOSX_LAI_ASSERT_EQ( src.localSize(), dst.localSize() );
+
+    real64 const * const src_data = src.extractLocalVector();
+    real64 * const dst_data = dst.extractLocalVector();
+    real64 const * const diagInv_data = m_diagInv->extractLocalVector();
+    dst.copy( src );
+    for( localIndex i = 0; i < src.localSize(); ++i )
+    {
+      dst_data[i] = src_data[i] * diagInv_data[i];
+    }
+  }
+
+private:
+
+  /// The diagonal of the matrix
+  Vector * m_diagInv;
```

Comment:
Using the object `Vector` instead of a pointer to a `Vector`, I have a problem here (using _hypre_ LAI):
https://github.com/GEOSX/GEOSX/blob/29f3572a04efcb164ca80bb29112467d051becb2/src/coreComponents/linearAlgebra/interfaces/hypre/HypreVector.cpp#L102-L109
I think that this operation is called here:
https://github.com/GEOSX/GEOSX/blob/29f3572a04efcb164ca80bb29112467d051becb2/src/coreComponents/physicsSolvers/multiphysics/LagrangianContactSolver.cpp#L1139

The problem is that `m_diagInv` is not ready.
I came up with the pointer to avoid this, but if there are better solutions, I would be glad to switch to one of them!

Reply:
About the _readiness_ check, I agree. Added.

Reply:
In this case you can avoid the copy constructor
```c++
tracPrecond = std::make_unique< PreconditionerJacobi< LAInterface > >();
```
as our preconditioner classes aren't copy-safe anyway and we should probably disable copy constructors in a separate cleanup PR. Alternatively, maybe the `ready` check should be relaxed in vector.

Reply:
And 
```c++
tracPrecond = std::make_unique< PreconditionerBlockJacobi< LAInterface > >( mechParams.dofsPerNode );
```
for the other one

Reply:
Thanks @klevzoff! Without copy constructor, it works.